File: solutions/cpp/0092.cpp
class Solution {
 public:
  ListNode* reverseBetween(ListNode* head, int m, int n) {
    if (!head || m == n) return head;

    ListNode dummy(0);
    dummy.next = head;
    ListNode* prev = &dummy;

    for (int i = 0; i < m - 1; ++i)
      prev = prev->next;
    ListNode* tail = prev->next;

    for (int i = 0; i < n - m; ++i) {
      ListNode* cache = prev->next;
      prev->next = tail->next;
      tail->next = tail->next->next;
      prev->next->next = cache;
    }

    return dummy.next;
  }
};

File: solutions/java/0092.java
class Solution {
  public ListNode reverseBetween(ListNode head, int m, int n) {
    if (head == null)
      return null;

    ListNode prev = null;
    ListNode curr = head;

    for (int i = 0; i < m - 1; ++i) {
      prev = curr;
      curr = curr.next;
    }

    ListNode conn = prev;
    ListNode tail = curr;

    for (int i = 0; i < n - m + 1; ++i) {
      ListNode next = curr.next;
      curr.next = prev;
      prev = curr;
      curr = next;
    }

    if (conn != null)
      conn.next = prev;
    else
      head = prev;
    tail.next = curr;

    return head;
  }
}